File: swiftly/Services/AdvertisementService.cs
using System;
using System.Collections.Generic;

namespace NotifyMessages;

/// Сервис показа рекламы: управление таймерами и вывод сообщений.
///
/// Фреймворка здесь нет: таймер и рассылка приходят делегатами из плагина,
/// поэтому класс одинаков в cssharp/ и swiftly/.
public sealed class AdvertisementService
{
    private readonly Config _config;
    private readonly ILogger _logger;

    /// (интервал в секундах, действие) -> как остановить
    private readonly Func<float, Action, Action> _repeatEvery;

    /// Рассылка всем игрокам в указанный канал
    private readonly Action<MessageType, string> _broadcast;

    private readonly List<Action> _stopTimers = new();

    public AdvertisementService(
        Config config,
        ILogger logger,
        Func<float, Action, Action> repeatEvery,
        Action<MessageType, string> broadcast)
    {
        _config = config;
        _logger = logger;
        _repeatEvery = repeatEvery;
        _broadcast = broadcast;
    }

    public void Start()
    {
        if (_config.Ads == null || _config.Ads.Count == 0)
            return;

        for (var i = 0; i < _config.Ads.Count; i++)
        {
            var ad = _config.Ads[i];

            if (ad.Messages == null || ad.Messages.Count == 0)
            {
                _logger.Info($"[ADS] Block #{i + 1} has no messages, skipped");
                continue;
            }

            // Запускаем цикл показа конкретного блока рекламы
            _stopTimers.Add(_repeatEvery(Math.Max(1f, ad.Interval), () => ShowAd(ad)));
        }
    }

    public void Stop()
    {
        foreach (var stop in _stopTimers) stop();
        _stopTimers.Clear();
    }

    private void ShowAd(Advertisement ad)
    {
        var messages = ad.NextMessages;
        if (messages == null) return;

        foreach (var (type, message) in messages)
        {
            // Одна нотация канала на весь конфиг. Раньше здесь был switch по точному регистру:
            // "chat" молча терялся, а Alert и CenterHtml не поддерживались вовсе.
            if (!Enum.TryParse<MessageType>(type, ignoreCase: true, out var channel))
            {
                // Это ошибка конфига, а не отладочный шум — админ должен её увидеть
                _logger.Info($"[ADS] Неизвестный канал '{type}'. Допустимые: " +
                             "Chat, Center, CenterHtml, Console, Alert");
                continue;
            }

            // Делегируем обработку/локализацию и фактический вывод наружу
            _broadcast(channel, message);
        }
    }
}

File: metamod/src/core/advertisement_service.cpp
#include "core/advertisement_service.h"

#include "core/logger.h"

#include <utility>

namespace nm {

AdvertisementService::AdvertisementService(const std::vector<Advertisement>& ads, ILogger* logger,
                                           RepeatEvery repeatEvery, Broadcast broadcast)
    : _ads(ads), _logger(logger), _repeatEvery(std::move(repeatEvery)), _broadcast(std::move(broadcast)) {}

void AdvertisementService::Start() {
    for (size_t i = 0; i < _ads.size(); ++i) {
        if (_ads[i].messages.empty()) {
            _logger->Info("[ADS] Block #" + std::to_string(i + 1) + " has no messages, skipped");
            continue;
        }

        const float interval = _ads[i].interval > 1.0f ? _ads[i].interval : 1.0f;
        _stopTimers.push_back(_repeatEvery(interval, [this, i]() { ShowAd(i); }));
    }
}

void AdvertisementService::Stop() {
    for (auto& stop : _stopTimers) {
        if (stop) stop();
    }
    _stopTimers.clear();
}

void AdvertisementService::ShowAd(size_t index) {
    const OrderedPairs* messages = _ads[index].NextMessages();
    if (messages == nullptr) return;

    for (const auto& message : *messages) {
        // Одна нотация канала на весь конфиг: "chat" и "Chat" — одно и то же
        MessageType channel = MessageType::Chat;
        if (!ParseMessageType(message.first, &channel)) {
            // Это ошибка конфига, а не отладочный шум — админ должен её увидеть
            _logger->Info("[ADS] Неизвестный канал '" + message.first +
                          "'. Допустимые: Chat, Center, CenterHtml, Console, Alert");
            continue;
        }
        _broadcast(channel, message.second);
    }
}

}  // namespace nm
